### ISSUE TITLE
UnicodeEncodeError: 'ascii' codec can't encode character u'\xf6' in position 4: ordinal not in range(128)

### ISSUE THREAD

**Issue by JensKorte** [pimutils/vdirsyncer#315]
Hi
I have problems in syncing to an existing owncloud. I tracked the problem to 

```
#CALDAV
[pair test]
a = test_calendar_local
b = test_calendar_remote
collections = ["persönlich"]
#collections = ["test"]
```

If I use `test` instead of `persönlich` the configuration works. If I replace `ö` with `%C3%B6` it doesn't work either.

I installed vdirsyncer with pip install and also tried `pip install -U git+https://github.com/untitaker/vdirsyncer.git`.

```
Python 2.7.3 (default, Mar 14 2014, 11:57:14)
vdirsyncer, version 0.7.6.dev1+ng99f5e51
```

Do you have any hint for me?

**Comment by untitaker** [comment]
Please post the full output

On 4 January 2016 11:38:07 CET, JensKorte <notifications@github.com> wrote:
>Hi
>I have problems in syncing to an existing owncloud I tracked the
>problem to 
>
>#CALDAV
>[pair test]
>a = test_calendar_local
>b = test_calendar_remote
>collections = ["persönlich"]
>#collections = ["test"]
>
>If I use test instead of persönlich the configuration works If I
>replace ö with %C3%B6 it doesn't work either
>
>I installed vdirsyncer with pip install and also tried pip install -U
>git+https://githubcom/untitaker/vdirsyncergit
>
>Python 273 (default, Mar 14 2014, 11:57:14)
>vdirsyncer, version 076dev1+ng99f5e51
>
>Do you have any hint for me?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/untitaker/vdirsyncer/issues/315

-- 
Sent from my phone. Please excuse my brevity.

**Comment by JensKorte** [comment]
vdirsyncer sync
Discovering collections for pair test
Unhandled exception occured.
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/utils.py", line 427, in _worker
    func(wq=self)
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/tasks.py", line 18, in prepare_pair
    status_path=config.general['status_path'], pair=pair
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/utils.py", line 173, in collections_for_pair
    rv = list(_collections_for_pair_impl(status_path, pair))
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/utils.py", line 272, in _collections_for_pair_impl
    collection)
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/utils.py", line 231, in _handle_collection_not_found
    collection, storage_name))
UnicodeEncodeError: 'ascii' codec can't encode character u'\xf6' in position 4: ordinal not in range(128)

.vdirsyncer/config:
[general]
# A folder where vdirsyncer can store some metadata about each pair.
status_path = ~/.vdirsyncer/status/

# CALDAV
[pair test]
a = test_calendar_local
b = test_calendar_remote
collections = ["persönlich"]
#collections = ["persönlich", "work"]

# Calendars also have a color property
metadata = ["displayname", "color"]

[storage test_calendar_local]
type = filesystem
path = ~/.calendars/
fileext = .ics

[storage test_calendar_remote]
type = caldav
url = http://IP/owncloud/remote.php/caldav/calendars/user/
username = xxx
password = xxx

**Comment by untitaker** [comment]
Ok, will take a look later. I suspect that Python 3 would work better.

**Comment by untitaker** [comment]
Okay, please try:

    pip install -U git+https://github.com/untitaker/vdirsyncer.git#issue315

**Comment by JensKorte** [comment]
Yesterday I tried some things, like python 3.2, but nothing worked.

 pip install -U git+https://github.com/untitaker/vdirsyncer.git#issue315
 Successfully installed vdirsyncer
 Cleaning up... 

vdirsyncer sync
Discovering collections for pair test
Unhandled exception occured.
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/utils.py", line 427, in _worker
    func(wq=self)
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/tasks.py", line 18, in prepare_pair
    status_path=config.general['status_path'], pair=pair
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/utils.py", line 173, in collections_for_pair
    rv = list(_collections_for_pair_impl(status_path, pair))
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/utils.py", line 272, in _collections_for_pair_impl
    collection)
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/utils.py", line 231, in _handle_collection_not_found
    collection, storage_name))
UnicodeEncodeError: 'ascii' codec can't encode character u'\xf6' in position 4: ordinal not in range(128)

python --version: Python 2.7.3
vdirsyncer  --version
vdirsyncer, version 0.7.6.dev1+ng99f5e51

**Comment by JensKorte** [comment]
@untitaker 
'''vdirsyncer  --version
vdirsyncer, version 0.7.6.dev1+ng99f5e51'''
Is this the master tree?

**Comment by untitaker** [comment]
Python 3.2 and 3.1 are not supported, only 3.3+.

Will have a look later.

**Comment by JensKorte** [comment]
It seems to me, that still the master tree is used, but I don't know how to switch to issue315 tree.
```
pip install -U git+https://github.com/untitaker/vdirsyncer.git#issue315
Downloading/unpacking git+https://github.com/untitaker/vdirsyncer.git
  Cloning https://github.com/untitaker/vdirsyncer.git to /tmp/pip-VB6Q_T-build
  Running setup.py egg_info for package from git+https://github.com/untitaker/vdirsyncer.git
    
    Installed /tmp/pip-VB6Q_T-build/.eggs/setuptools_scm-1.10.1-py2.7.egg

    no previously-included directories found matching 'docs/_build'
    warning: no previously-included files matching '*.py[cdo]' found anywhere in distribution
    warning: no previously-included files matching '__pycache__' found anywhere in distribution
    warning: no previously-included files matching '*.so' found anywhere in distribution
    warning: no previously-included files matching '*.pyd' found anywhere in distribution
Requirement already up-to-date: click>=5.0 in /usr/local/lib/python2.7/dist-packages (from vdirsyncer==0.7.6.dev1+ng99f5e51)
Requirement already up-to-date: click-log in /usr/local/lib/python2.7/dist-packages (from vdirsyncer==0.7.6.dev1+ng99f5e51)
Requirement already up-to-date: click-threading in /usr/local/lib/python2.7/dist-packages (from vdirsyncer==0.7.6.dev1+ng99f5e51)
Requirement already up-to-date: requests!=2.9.0 in /usr/local/lib/python2.7/dist-packages (from vdirsyncer==0.7.6.dev1+ng99f5e51)
Requirement already up-to-date: lxml>=3.1 in /usr/local/lib/python2.7/dist-packages (from vdirsyncer==0.7.6.dev1+ng99f5e51)
Requirement already up-to-date: requests-toolbelt>=0.5.0 in /usr/local/lib/python2.7/dist-packages (from vdirsyncer==0.7.6.dev1+ng99f5e51)
Requirement already up-to-date: atomicwrites in /usr/local/lib/python2.7/dist-packages (from vdirsyncer==0.7.6.dev1+ng99f5e51)
Installing collected packages: vdirsyncer
  Found existing installation: vdirsyncer 0.7.5
    Uninstalling vdirsyncer:
      Successfully uninstalled vdirsyncer
  Running setup.py install for vdirsyncer

    no previously-included directories found matching 'docs/_build'
    warning: no previously-included files matching '*.py[cdo]' found anywhere in distribution
    warning: no previously-included files matching '__pycache__' found anywhere in distribution
    warning: no previously-included files matching '*.so' found anywhere in distribution
    warning: no previously-included files matching '*.pyd' found anywhere in distribution
    Installing vdirsyncer script to /usr/local/bin
Successfully installed vdirsyncer
Cleaning up...
```

**Comment by untitaker** [comment]
Argh, replace # with @

On 8 January 2016 18:20:46 CET, JensKorte <notifications@github.com> wrote:
>It seems to me, that still the master tree is used, but I don't know
>how to switch to issue315 tree.
>```
>pip install -U git+https://github.com/untitaker/vdirsyncer.git#issue315
>Downloading/unpacking git+https://github.com/untitaker/vdirsyncer.git
>Cloning https://github.com/untitaker/vdirsyncer.git to
>/tmp/pip-VB6Q_T-build
>Running setup.py egg_info for package from
>git+https://github.com/untitaker/vdirsyncer.git
>    
>  Installed /tmp/pip-VB6Q_T-build/.eggs/setuptools_scm-1.10.1-py2.7.egg
>
>    no previously-included directories found matching 'docs/_build'
>warning: no previously-included files matching '*.py[cdo]' found
>anywhere in distribution
>warning: no previously-included files matching '__pycache__' found
>anywhere in distribution
>warning: no previously-included files matching '*.so' found anywhere in
>distribution
>warning: no previously-included files matching '*.pyd' found anywhere
>in distribution
>Requirement already up-to-date: click>=5.0 in
>/usr/local/lib/python2.7/dist-packages (from
>vdirsyncer==0.7.6.dev1+ng99f5e51)
>Requirement already up-to-date: click-log in
>/usr/local/lib/python2.7/dist-packages (from
>vdirsyncer==0.7.6.dev1+ng99f5e51)
>Requirement already up-to-date: click-threading in
>/usr/local/lib/python2.7/dist-packages (from
>vdirsyncer==0.7.6.dev1+ng99f5e51)
>Requirement already up-to-date: requests!=2.9.0 in
>/usr/local/lib/python2.7/dist-packages (from
>vdirsyncer==0.7.6.dev1+ng99f5e51)
>Requirement already up-to-date: lxml>=3.1 in
>/usr/local/lib/python2.7/dist-packages (from
>vdirsyncer==0.7.6.dev1+ng99f5e51)
>Requirement already up-to-date: requests-toolbelt>=0.5.0 in
>/usr/local/lib/python2.7/dist-packages (from
>vdirsyncer==0.7.6.dev1+ng99f5e51)
>Requirement already up-to-date: atomicwrites in
>/usr/local/lib/python2.7/dist-packages (from
>vdirsyncer==0.7.6.dev1+ng99f5e51)
>Installing collected packages: vdirsyncer
>  Found existing installation: vdirsyncer 0.7.5
>    Uninstalling vdirsyncer:
>      Successfully uninstalled vdirsyncer
>  Running setup.py install for vdirsyncer
>
>    no previously-included directories found matching 'docs/_build'
>warning: no previously-included files matching '*.py[cdo]' found
>anywhere in distribution
>warning: no previously-included files matching '__pycache__' found
>anywhere in distribution
>warning: no previously-included files matching '*.so' found anywhere in
>distribution
>warning: no previously-included files matching '*.pyd' found anywhere
>in distribution
>    Installing vdirsyncer script to /usr/local/bin
>Successfully installed vdirsyncer
>Cleaning up...
>```
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/untitaker/vdirsyncer/issues/315#issuecomment-170063145

-- 
Sent from my phone. Please excuse my brevity.

**Comment by JensKorte** [comment]
```
vdirsyncer --version
vdirsyncer, version 0.7.6.dev2+ng8e4edb5

vdirsyncer sync
Discovering collections for pair test
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/click_log/core.py", line 42, in emit
    msg = self.format(record)
  File "/usr/lib/python2.7/logging/__init__.py", line 723, in format
    return fmt.format(record)
  File "/usr/local/lib/python2.7/dist-packages/click_log/core.py", line 34, in forma
t
    for x in str(record.msg).splitlines())
UnicodeEncodeError: 'ascii' codec can't encode character u'\xf6' in position 18: ord
inal not in range(128)
Logged from file utils.py, line 231
Should vdirsyncer attempt to create it? [y/N]: y
Unhandled exception occured.
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/utils.py", line 427, i
n _worker
    func(wq=self)
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/tasks.py", line 61, in
 sync_collection
    handle_cli_error(status_name)
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/utils.py", line 122, i
n handle_cli_error
    msg = 'Unhandled exception occured for {}.'.format(status_name)
UnicodeEncodeError: 'ascii' codec can't encode character u'\xf6' in position 9: ordi
nal not in range(128)
```
running again, without the question "Should vdirsyncer attempt to create it? [y/N]: y"
```
vdirsyncer sync
Unhandled exception occured.
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/utils.py", line 427, in _worker
    func(wq=self)
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/tasks.py", line 61, in sync_collection
    handle_cli_error(status_name)
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/utils.py", line 122, in handle_cli_error
    msg = 'Unhandled exception occured for {}.'.format(status_name)
UnicodeEncodeError: 'ascii' codec can't encode character u'\xf6' in position 9: ordinal not in range(128)
```

**Comment by untitaker** [comment]
So that should be fixed now, please try the pip-install command again...

BTW, see #219 for deprecation of Python 2 support, Python 3 is recommended.

**Comment by JensKorte** [comment]
I will have a look at Python 3.3
```
vdirsyncer sync
Unhandled exception occured.
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/utils.py", line 427, i
n _worker
    func(wq=self)
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/tasks.py", line 61, in
 sync_collection
    handle_cli_error(status_name)
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/utils.py", line 122, i
n handle_cli_error
    msg = 'Unhandled exception occured for {}.'.format(status_name)
UnicodeEncodeError: 'ascii' codec can't encode character u'\xf6' in position 9: ordi
nal not in range(128)

vdirsyncer --version
vdirsyncer, version 0.7.6.dev2+ngbb3ce2d
```

**Comment by untitaker** [comment]
That's the wrong version. Please try `pip uninstall vdirsyncer`, followed by the install command again.

**Comment by JensKorte** [comment]
I did ```pip uninstall vdirsyncer``` but /usr/local/bin/vdirsyncer was still there. I removed it manually (rm).
```
pip install -U git+https://github.com/untitaker/vdirsyncer.git@issue315
vdirsyncer --version
vdirsyncer, version 0.7.6.dev2+ngbb3ce2d
```

**Comment by untitaker** [comment]
NOW everything really should be fixed. I've even added tests!

**Comment by JensKorte** [comment]
I am prepared to install python 3.3, but I don't want to change the system while we are bug hunting. 

```
vdirsyncer --version
vdirsyncer, version 0.7.6.dev6+ng650b749
$ vdirsyncer sync
Syncing test/persönlich

Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/utils.py", line 356, in storage_instance_from_config
    return cls(**new_config)
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/storage/filesystem.py", line 44, in __init__
    super(FilesystemStorage, self).__init__(**kwargs)
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/storage/base.py", line 80, in __init__
    instance_name = '{}/{}'.format(instance_name, collection)
UnicodeEncodeError: 'ascii' codec can't encode character u'\xf6' in position 4: ordinal not in range(128)
critical: Failed to initialize test_calendar_local
critical:   - 'ascii' codec can't encode character u'\xf6' in position 4: ordinal not in range(128)
```

**Comment by untitaker** [comment]
I see. It fails on my machine too, but the fix is painful for Python 2. I've pushed a workaround that will last as long as Python 2 support does.

Please use `pip uninstall vdirsyncer && pip install -U git+https://github.com/untitaker/vdirsyncer.git`

**Comment by JensKorte** [comment]
```
vdirsyncer --version
vdirsyncer, version 0.7.6.dev7+ng7646222

$ vdirsyncer sync
Syncing test/persönlich
Unhandled exception occured for test/persönlich.
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/cli/tasks.py", line 61, in sync_collection
    force_delete=force_delete
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/sync.py", line 221, in sync
    action(a_info, b_info, conflict_resolution)
  File "/usr/local/lib/python2.7/dist-packages/vdirsyncer/sync.py", line 232, in inner
    .format(ident, dest.storage))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 24: ordinal not in range(128)
```

**Comment by JensKorte** [comment]
I added the stable tree of debian to my repositories and blocked it with pinning, oldstable is preferred. Now I am able to install python 3.4.

**Comment by untitaker** [comment]
This now really should be fixed in master, but TBH I don't care enough to fix Python 2 support.

Apart from that I am not sure whether your collection really is called `persönlich`. Which server is this that doesn't remove non-ASCII characters from their URLs?

**Comment by untitaker** [comment]
@JensKorte Am I correct in assuming that everything works for your with Python 3?

**Comment by untitaker** [comment]
Closing because it'll be fixed with #219
